### PR TITLE
Chri o neill45

### PR DESCRIPTION
**Pull request by PerryBaran**

### CODE REVIEW

File: src/kata1.fizzBuzz.js [hunk 1]
```diff
@@ -1,3 +1,18 @@
...
+    return "Buzz";
+  }
+  if (isDivisableBy3) {
+    return "Fizz";
+  }
+
+  return number;
+};
 
 module.exports = fizzBuzz;
```
Comment (PerryBaran) [inline src/kata1.fizzBuzz.js:18]: Good

File: test/kata1.fizzBuzz.test.js [hunk 2]
```diff
@@ -1,11 +1,22 @@
...
+    expect(fizzBuzz(15)).toBe("FizzBuzz");
+    expect(fizzBuzz(30)).toBe("FizzBuzz");
+    expect(fizzBuzz(90)).toBe("FizzBuzz");
+  });
 
-  test("returns the number when it isn't a multiple of 3 or 5", () => {});
+  it("returns the number when it isn't a multiple of 3 or 5", () => {
+    expect(fizzBuzz(4)).toBe(4);
+    expect(fizzBuzz(11)).toBe(11);
+  });
```
Comment (PerryBaran) [inline test/kata1.fizzBuzz.test.js:21]: good tests

File: src/kata2.booleanToWord.js [hunk 3]
```diff
@@ -1,3 +1,2 @@
-const booleanToWord = (boolean) => {};
-
+const booleanToWord = (boolean) => (boolean ? "Yes" : "No");
 module.exports = booleanToWord;
```
Comment (PerryBaran) [inline src/kata2.booleanToWord.js:2]: Good, clean solution. You should leave an empty line before `module.exports`

File: test/kata2.booleanToWord.test.js [hunk 4]
```diff
@@ -1,5 +1,11 @@
...
 
 describe("booleanToWord", () => {
-  // how do we create specs again???
+  it("should return 'Yes' when input is true", () => {
+    expect(booleanToWord(true)).toBe("Yes");
+  });
+
+  it("should return 'No' when input is falso", () => {
+    expect(booleanToWord(false)).toBe("No");
+  });
```
Comment (PerryBaran) [inline test/kata2.booleanToWord.test.js:10]: Good

File: src/kata3.numberToReversedDigits.js [hunk 5]
```diff
@@ -1,3 +1,8 @@
-const numberToReversedDigits = (number) => {};
+const numberToReversedDigits = (number) => {
+  const newArray = String(number).split("").map(Number);
+  const newReversedArray = newArray.reverse();
+
+  return newReversedArray;
```
Comment (PerryBaran) [inline src/kata3.numberToReversedDigits.js:5]: Good solution

File: test/kata3.numberToReversedDigits.test.js [hunk 6]
```diff
@@ -1,5 +1,7 @@
 const { numberToReversedDigits } = require("../src");
 
 describe("numberToReversedDigits", () => {
-  test("returns a reversed array of the number's digits", () => {});
-});
\ No newline at end of file
+  test("returns a reversed array of the number's digits", () => {
+    expect(numberToReversedDigits(12345)).toEqual([5, 4, 3, 2, 1]);
+  });
```
Comment (PerryBaran) [inline test/kata3.numberToReversedDigits.test.js:6]: Probably be a good idea to have more expect statements with different inputs just to confirm it works

File: test/kata4.humanCatDogYears.test.js [hunk 7]
```diff
@@ -1,3 +1,9 @@
 const { humanCatDogYears } = require("../src");
 
-// Look Ma, no handlebars!!!
+describe("humanCatDogYears", () => {
+  it("turns human years into cat and dog years", () => {
+    expect(humanCatDogYears(1)).toEqual([1, 15, 15]);
+    expect(humanCatDogYears(2)).toEqual([2, 24, 24]);
+    expect(humanCatDogYears(10)).toEqual([10, 56, 64]);
+  });
```
Comment (PerryBaran) [inline test/kata4.humanCatDogYears.test.js:8]: I would separate out these tests into there own blocks for each edge case. I would also add a test for if human years is 0 e.g.
```
it("turns human years into cat and dog years", () => {
  ...
});

it("converts 1 human year into cat and dog years", () => {
  ...
});

it("converts 2 human years into cat and dog years", () => {
  ...
});

it("converts 0 human year into cat and dog years", () => {
  ...
});
```

File: src/kata4.humanCatDogYears.js [hunk 8]
```diff
@@ -1,3 +1,18 @@
-const humanCatDogYears = (number) => {};
+const humanCatDogYears = (number) => {
+  let catYears;
+  let dogYears;
```
Comment (PerryBaran) [inline src/kata4.humanCatDogYears.js:3]: When creating variables you usually assign them a value of the type you want them to be, otherwise they will take the value of `undefined` which could lead to unexpected problems later. e.g.
```
let catYears = 0;
let dogYears = 0;
```

File: src/kata4.humanCatDogYears.js [hunk 9]
```diff
@@ -1,3 +1,18 @@
...
+    dogYears = 15;
+  } else if (number === 2) {
+    catYears = 24;
+    dogYears = 24;
+  } else {
+    catYears = 24 + (number - 2) * 4;
+    dogYears = 24 + (number - 2) * 5;
+  }
+  return [number, catYears, dogYears];
+};
```
Comment (PerryBaran) [inline src/kata4.humanCatDogYears.js:16]: If you initialise `catYears` and `dogYears` with `0`, you can then have a series of `if` statements that sequentially add onto this value with increasing `number`. This will reduce code repetition and will make this function work for `0` human years. i.e.
```
const humanCatDogYears = (number) => {
  let catYears = 0;
  let dogYears = 0;

  if (number >= 1) {
    catYears += 15;
    dogYears += 15;
  }

  if (number >= 2) {
    catYears += 9;
    dogYears += 9;
  }

  if (number > 2) {
    catYears += (number - 2) * 4;
    dogYears += (number - 2) * 5;
  }

  return [number, catYears, dogYears];
};
```

File: src/kata5.reachDestination.js [hunk 10]
```diff
@@ -1,3 +1,6 @@
-const reachDestination = (distance, speed) => {};
+const reachDestination = (distance, speed) => {
+  const arrivalTime = Math.ceil((distance / speed) * 2) / 2;
+  return `I will be there in ${arrivalTime}, see you soon.`;
```
Comment (PerryBaran) [inline src/kata5.reachDestination.js:3]: Be there in how long? hours? minutes? years? You should include these units in the string. e.g.
```
 return `I will be there in ${arrivalTime} hours, see you soon.`;
```
However, then if `arrivalTime` is `1` your sentence would then be grammatically incorrect `"I will be there in 1 hours, see you soon."` instead of `"I will be there in 1 hour, see you soon."`. To fix this, you can conditionally add the `"s"` at the end of `"hour"` e.g.
```
return `I will be there in ${arrivalTime} hour${arrivalTime === 1 ? "" : "s"}, see you soon.`;
```

File: test/kata5.reachDestination.test.js [hunk 11]
```diff
@@ -1,5 +1,9 @@
 const { reachDestination } = require("../src");
 
 describe("reachDestination", () => {
-  test("returns string with estimated time of arrival", () => {});
+  test("returns string with estimated time of arrival", () => {
+    expect(reachDestination(44, 10)).toEqual(
+      "I will be there in 4.5, see you soon."
+    );
+  });
```
Comment (PerryBaran) [inline test/kata5.reachDestination.test.js:8]: As previously mentioned, should add the units for time to this string. After doing this, I would add a separate test for your edge case where the arrival time is `1` e.g.
```
test("returns a string with the estimated time of arrival for 1 hour", () => {
    expect(reachDestination(1, 1)).toEqual("I will be there in 1 hour, see you soon.");
  });
```

File: src/kata6.joinNames.js [hunk 12]
```diff
@@ -1,3 +1,11 @@
-const joinNames = (namesObj) => {};
+const joinNames = (namesObj) => {
+  const nameArray = namesObj.map((item) => item.name);
+  if (nameArray.length <= 1) {
+    return nameArray.join("");
+  } else {
+    const lastName = nameArray.pop();
+    return nameArray.join(", ") + " & " + lastName;
+  }
```
Comment (PerryBaran) [inline src/kata6.joinNames.js:8]: unnecessary `else`, if your `if` statement contains a `return` you do not need the `else` keyword (as, if the `return` is ran, not code after that in the function will be run) i.e.
```
const joinNames = (namesObj) => {
  const nameArray = namesObj.map((item) => item.name);
  if (nameArray.length <= 1) {
    return nameArray.join("");
  }
  const lastName = nameArray.pop();
  return nameArray.join(", ") + " & " + lastName;
};
```

File: src/kata6.joinNames.js [hunk 13]
```diff
@@ -1,3 +1,11 @@
-const joinNames = (namesObj) => {};
+const joinNames = (namesObj) => {
+  const nameArray = namesObj.map((item) => item.name);
+  if (nameArray.length <= 1) {
+    return nameArray.join("");
+  } else {
+    const lastName = nameArray.pop();
+    return nameArray.join(", ") + " & " + lastName;
```
Comment (PerryBaran) [inline src/kata6.joinNames.js:7]: string concatenation is preferred by the linter, i.e.
```
return `${nameArray.join(", ")} & ${lastName}`;
```

File: test/kata6.joinNames.test.js [hunk 14]
```diff
@@ -1,5 +1,8 @@
 const { joinNames } = require("../src");
 
 describe("joinNames", () => {
-  test("returns string of names, seperated by commas and an ampersand", () => {});
+  test("returns string of names, seperated by commas and an ampersand", () => {
+    const inputNames = [{ name: "Bart" }, { name: "Lisa" }, { name: "Maggie" }];
+    expect(joinNames(inputNames)).toEqual("Bart, Lisa & Maggie");
+  });
```
Comment (PerryBaran) [inline test/kata6.joinNames.test.js:7]: Should add more tests for edge cases. I would have separate test for when `inputNames` has a length of: 0, 1, 2, 3 and more than 3.

I have tested for each of these scenarios and your code works as expected. You should still create these test to confirm that and in case someone wants to go back and alter this function later, they can then be sure that the functionality hasn't been broken and get a better understanding of how this function should work with different inputs.

File: src/kata7.getEmployerRole.js [hunk 15]
```diff
@@ -1,3 +1,9 @@
-const getEmployerRole = (employeeName, employees) => {};
+const getEmployerRole = (employeeName, employees) => {
+  for (const employee of employees) {
+    if (employee.name === employeeName) {
+      return employee.role;
+    }
+  }
```
Comment (PerryBaran) [inline src/kata7.getEmployerRole.js:6]: Good solution, you could also do create this functionality using `.find`, e.g.
```
return employees.find(employee => employee.name === employeeName).role;
```

**Review by PerryBaran — COMMENTED** [review]
Overall really good. Some minor formatting issues + Could use more tests for edge cases. I've left comments throughout where I think it could be improved.

File: test/kata4.humanCatDogYears.test.js (hunk 7 above)
Comment (Chris-ONeill45) [reply test/kata4.humanCatDogYears.test.js:8]: I will amend, thanks for the input.

File: src/kata4.humanCatDogYears.js (hunk 8 above)
Comment (Chris-ONeill45) [reply src/kata4.humanCatDogYears.js:3]: I'll amend, and that's a good tip which i'll always employ. Thanks!

File: src/kata5.reachDestination.js (hunk 10 above)
Comment (Chris-ONeill45) [reply src/kata5.reachDestination.js:3]: Wow, that's incredible! My mistake, i just missed th hours from my string. I can see how easy it is to miss something so simple.
